File: Assets/Bundles/Shared/Resources/TS/Item/HeldItems/BlockPlacement/BlockHeldItem.lua
-- Compiled with unity-ts v2.1.0-75
local Flamework = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Flamework
local WorldAPI = require("Shared/TS/VoxelWorld/WorldAPI").WorldAPI
local HeldItem = require("Shared/TS/Item/HeldItems/HeldItem").HeldItem
-- Dependencies
local BlockHeldItem
do
	local super = HeldItem
	BlockHeldItem = setmetatable({}, {
		__tostring = function()
			return "BlockHeldItem"
		end,
		__index = super,
	})
	BlockHeldItem.__index = BlockHeldItem
	function BlockHeldItem.new(...)
		local self = setmetatable({}, BlockHeldItem)
		return self:constructor(...) or self
	end
	function BlockHeldItem:constructor(...)
		super.constructor(self, ...)
		self.characterLayerMask = LayerMask:GetMask("Character")
	end
	function BlockHeldItem:OnEquip()
		super.OnEquip(self)
		if self.entity:IsLocalCharacter() then
			(Flamework.resolveDependency("Bundles/Client/Controllers/Global/BlockInteractions/BlockSelectController@BlockSelectController")):Enable()
		end
	end
	function BlockHeldItem:OnUnEquip()
		super.OnUnEquip(self)
		if self.entity:IsLocalCharacter() then
			(Flamework.resolveDependency("Bundles/Client/Controllers/Global/BlockInteractions/BlockSelectController@BlockSelectController")):Disable()
		end
	end
	function BlockHeldItem:OnUseClient(useIndex)
		-- Only run for local player
		if self.entity:IsLocalCharacter() then
			-- Try to place a block
			if self:TryPlaceBlock() then
				-- Only play use animations if we actually think we can place a block
				super.OnUseClient(self, useIndex)
			end
		end
	end
	function BlockHeldItem:TryPlaceBlock()
		if not self.meta.block then
			return false
		end
		local blockSelectController = Flamework.resolveDependency("Bundles/Client/Controllers/Global/BlockInteractions/BlockSelectController@BlockSelectController")
		local placePosition = blockSelectController.PlaceBlockPosition
		local isVoidPlacement = blockSelectController.IsVoidPlacement
		if not placePosition then
			return false
		end
		-- Make sure this position is valid within the playable area
		if (Flamework.resolveDependency("Bundles/Client/Controllers/Global/BlockInteractions/DenyRegionController@DenyRegionController")):InDenyRegion(placePosition) then
			return false
		end
		-- Prevent placing in an entity's head
		-- const collider = this.entity.references.characterCollider;
		-- const bounds = collider.bounds;
		-- const size = bounds.size;
		local _fn = Physics
		local _vector3 = Vector3.new(0.5, 0.5, 0.5)
		local colliders = _fn:OverlapBox(placePosition + _vector3, Vector3.new(0.5, 0.5, 0.5), Quaternion.identity, self.characterLayerMask)
		do
			local i = 0
			local _shouldIncrement = false
			while true do
				if _shouldIncrement then
					i += 1
				else
					_shouldIncrement = true
				end
				if not (i < colliders.Length) then
					break
				end
				local collider = colliders:GetValue(i)
				local center = collider.bounds.center
				if placePosition.y + 0.5 > center.y then
					return false
				end
			end
		end
		-- Write the voxel at the predicted position
		WorldAPI:GetMainWorld():PlaceBlockById(placePosition, self.meta.block.blockId, {
			placedByEntityId = self.entity.id,
		});
		(Flamework.resolveDependency("Bundles/Client/Controllers/Global/Character/LocalEntityController@LocalEntityController")):AddToMoveData("PlaceBlock", {
			pos = placePosition,
			itemType = self.meta.ItemType,
		})
		if isVoidPlacement then
			blockSelectController:PlacedVoidBridgeBlock()
		end
		return true
	end
end
return {
	BlockHeldItem = BlockHeldItem,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Shared/Item/HeldItems/BlockPlacement/BlockHeldItem.ts
import { Dependency } from "@easy-games/flamework-core";
import { WorldAPI } from "../../../VoxelWorld/WorldAPI";
import { HeldItem } from "../HeldItem";
//Dependencies
import { BlockSelectController } from "../../../../Client/Controllers/Global/BlockInteractions/BlockSelectController";
import { DenyRegionController } from "../../../../Client/Controllers/Global/BlockInteractions/DenyRegionController";
import { LocalEntityController } from "../../../../Client/Controllers/Global/Character/LocalEntityController";
import { Network } from "../../../Network";
import { Block } from "../../../VoxelWorld/Block";
import { Entity } from "../../../Entity/Entity";

export class BlockHeldItem extends HeldItem {
	private characterLayerMask = LayerMask.GetMask("Character");

	override OnEquip() {
		super.OnEquip();
		if (this.entity.IsLocalCharacter()) {
			Dependency<BlockSelectController>().Enable();
		}

		//Load the blocks mesh
		if (this.meta.block?.blockId) {
			const blockDefinition = WorldAPI.GetMainWorld().GetBlockDefinition(this.meta.block.blockId);
			const blockGO = MeshProcessor.ProduceSingleBlock(
				this.meta.block.blockId,
				WorldAPI.GetMainWorld().voxelWorld,
			);
			const gameObjects = this.entity.accessoryBuilder.GetAccessories(AccessorySlot.RightHand);
			blockGO.transform.SetParent(gameObjects.GetValue(0).transform);
			blockGO.transform.localPosition = new Vector3(0, 0, 0);
			const scale = 1;
			blockGO.transform.localScale = new Vector3(scale, scale, scale);
			blockGO.transform.localRotation = Quaternion.identity;
			blockGO.transform.Rotate(new Vector3(90, 90, 0));
		}
	}

	override OnUnEquip() {
		super.OnUnEquip();
		if (this.entity.IsLocalCharacter()) {
			Dependency<BlockSelectController>().Disable();
		}
	}

	override OnUseClient(useIndex: number) {
		//Only run for local player
		if (this.entity.IsLocalCharacter()) {
			//Try to place a block
			if (this.TryPlaceBlock()) {
				//Only play use animations if we actually think we can place a block
				super.OnUseClient(useIndex);
			}
		}
	}

	private TryPlaceBlock(): boolean {
		if (!this.meta.block) {
			return false;
		}

		const blockSelectController = Dependency<BlockSelectController>();
		const placePosition = blockSelectController.PlaceBlockPosition;
		const isVoidPlacement = blockSelectController.IsVoidPlacement;
		if (!placePosition) {
			return false;
		}

		// Make sure this position is valid within the playable area
		if (Dependency<DenyRegionController>().InDenyRegion(placePosition)) {
			return false;
		}

		// Prevent placing in an entity's head
		// const collider = this.entity.references.characterCollider;
		// const bounds = collider.bounds;
		// const size = bounds.size;
		const colliders = Physics.OverlapBox(
			placePosition.add(new Vector3(0.5, 0.5, 0.5)),
			new Vector3(0.5, 0.5, 0.5),
			Quaternion.identity,
			this.characterLayerMask,
		);
		for (let i = 0; i < colliders.Length; i++) {
			const collider = colliders.GetValue(i);
			const center = collider.bounds.center;
			if (placePosition.y + 0.5 > center.y) {
				return false;
			}
		}

		// Write the voxel at the predicted position
		WorldAPI.GetMainWorld().PlaceBlockById(placePosition, this.meta.block.blockId!, {
			placedByEntityId: this.entity.id,
		});

		Dependency<LocalEntityController>().AddToMoveData("PlaceBlock", {
			pos: placePosition,
			itemType: this.meta.ItemType,
		});
		if (isVoidPlacement) {
			blockSelectController.PlacedVoidBridgeBlock();
		}
		return true;
	}
}
